test/shared/TestConnection.cs

```diff
@@ -157,11 +139,9 @@ public Task WaitForConnectionClose()
 
         private void ReceiveAsyncCompleted(object sender, SocketAsyncEventArgs e)
         {
-            if (e.BytesTransferred == 0)
-            {
-                var tcs = (TaskCompletionSource<object>)e.UserToken;
-                tcs.SetResult(null);
-            }
+            Assert.Equal(0, e.BytesTransferred);
```

Comment:
Isn't this normally called on some random thread in the `SocketAsyncEventArgs.Completed` callback meaning the actual test will just time out if the assertion fails. I think I prefer calling `tcs.SetException()`.